### PR TITLE
Fix 3756

### LINKED ISSUE #3756: after disrupt_kill_scylla scylla-manager-agent can fail on startup and never recover
test_id: d590d632-9008-4054-b332-78a432790fbe

Logs:
db-cluster - https://cloudius-jenkins-test.s3.amazonaws.com/d590d632-9008-4054-b332-78a432790fbe/20210628_152305/db-cluster-d590d632.tar.gz  |
loader-set - https://cloudius-jenkins-test.s3.amazonaws.com/d590d632-9008-4054-b332-78a432790fbe/20210628_152305/loader-set-d590d632.tar.gz  |
sct-runner - https://cloudius-jenkins-test.s3.amazonaws.com/d590d632-9008-4054-b332-78a432790fbe/20210628_152305/sct-runner-d590d632.tar.gz

Error message:

```
2021-06-28T11:26:54+00:00  longevity-100gb-4h-master-db-node-d590d632-4 !INFO    | scylla-manager-agent[52424]: {"L":"ERROR","T":"2021-06-28T11:26:54.596Z","M":"Bye","error":"server init: Get \"http://0.0.0.0:10000/v2/config/listen_address\": net/http: request canceled (Client.Timeout exceeded while awaiting headers)","_trace_id":"xv-YlV1GQoeAWkFoJd6nmA","errorStack":"main.glob..func6\n\tgithub.com/scylladb/scylla-manager/pkg/cmd/agent/root.go:66\ngithub.com/spf13/cobra.(*Command).execute\n\tgithub.com/spf13/cobra@v1.1.1/command.go:850\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tgithub.com/spf13/cobra@v1.1.1/command.go:958\ngithub.com/spf13/cobra.(*Command).Execute\n\tgithub.com/spf13/cobra@v1.1.1/command.go:895\nmain.main\n\tgithub.com/scylladb/scylla-manager/pkg/cmd/agent/main.go:11\nruntime.main\n\truntime/proc.go:225\nruntime.goexit\n\truntime/asm_amd64.s:1371\n","S":"github.com/scylladb/go-log.Logger.log\n\tgithub.com/scylladb/go-log@v0.0.6/logger.go:101\ngithub.com/scylladb/go-log.Logger.Error\n\tgithub.com/scylladb/go-log@v0.0.6/logger.go:84\nmain.glob..func6.1\n\tgithub.com/scylladb/scylla-manager/pkg/cmd/agent/root.go:56\nmain.glob..func6\n\tgithub.com/scylladb/scylla-manager/pkg/cmd/agent/root.go:66\ngithub.com/spf13/cobra.(*Command).execute\n\tgithub.com/spf13/cobra@v1.1.1/command.go:850\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tgithub.com/spf13/cobra@v1.1.1/command.go:958\ngithub.com/spf13/cobra.(*Command).Execute\n\tgithub.com/spf13/cobra@v1.1.1/command.go:895\nmain.main\n\tgithub.com/scylladb/scylla-manager/pkg/cmd/agent/main.go:11\nruntime.main\n\truntime/proc.go:225"}
2021-06-28T11:26:54+00:00  longevity-100gb-4h-master-db-node-d590d632-4 !INFO    | scylla-manager-agent[52424]: STARTUP ERROR: server init: Get "http://0.0.0.0:10000/v2/config/listen_address": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

### PR DESCRIPTION
**Pull request by dkropachev**
Fixes #3756 

```
Workaround for https://github.com/scylladb/scylla-manager/issues/2813
When scylla take too long time to bring api port up
scylla-manager-agent fails to start and never go up
```

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: sdcm/nemesis.py [hunk 1]
```diff
@@ -389,11 +389,16 @@ def _kill_scylla_daemon(self):
...
+                # Workaround for https://github.com/scylladb/scylla-manager/issues/2813
+                # When scylla take too long time to bring api port up
+                #  scylla-manager-agent fails to start and never go up
+                self.target_node.start_service('scylla-manager-agent', timeout=600, ignore_status=True)
             self.log.info('Waiting JMX services to be restarted after we killed them...')
             self.target_node.wait_jmx_up()
         self.cluster.wait_for_schema_agreement()
 
-    def disrupt_stop_wait_start_scylla_server(self, sleep_time=300):  # pylint: disable=invalid-name
+    def disrupt_sto4p_wait_start_scylla_server(self, sleep_time=300):  # pylint: disable=invalid-name
```
Comment (fgelcer) [inline sdcm/nemesis.py:401]: you introduced a type in the function name

**Review by fgelcer — CHANGES_REQUESTED** [review]

File: sdcm/cluster.py [hunk 2]
```diff
@@ -2377,23 +2377,38 @@ def add_install_prefix(self, abs_path):
...
+        else:
+            self.remoter.run(f'{self.systemctl} stop {service_name}.service', **opts)
+
+    def restart_service(self, service_name: str, timeout=500, ignore_status=False):
+        opts = {'timeout': timeout, 'ignore_status': ignore_status}
+        if self.is_ubuntu14():
+            self.remoter.run(f'sudo service {service_name} stop', **opts)
+        else:
+            self.remoter.run(f'{self.systemctl} stop {service_name}.service', **opts)
+
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/cluster.py:2400]: fixed

File: sdcm/nemesis.py (hunk 1 above)
Comment (dkropachev) [reply sdcm/nemesis.py]: fixed

File: sdcm/cluster.py [hunk 3]
```diff
@@ -2377,23 +2377,34 @@ def add_install_prefix(self, abs_path):
         }
         return checklist.get(abs_path, INSTALL_DIR + abs_path)
 
+    def _service_cmd(self, service_name: str, cmd: str, timeout: int = 500, ignore_status=False):
+        opts = {'timeout': timeout, 'ignore_status': ignore_status}
+        if self.is_ubuntu14():
+            cmd = f'sudo service {service_name} {cmd}'
+        else:
+            cmd = f'{self.systemctl} {cmd} {service_name}.service'
+        self.remoter.run(cmd, **opts)
```
Comment (bentsi) [inline sdcm/cluster.py:2386]: let's make it sudo

File: sdcm/cluster.py [hunk 4]
```diff
@@ -2377,23 +2377,34 @@ def add_install_prefix(self, abs_path):
         }
         return checklist.get(abs_path, INSTALL_DIR + abs_path)
 
+    def _service_cmd(self, service_name: str, cmd: str, timeout: int = 500, ignore_status=False):
+        opts = {'timeout': timeout, 'ignore_status': ignore_status}
+        if self.is_ubuntu14():
+            cmd = f'sudo service {service_name} {cmd}'
```
Comment (fgelcer) [inline sdcm/cluster.py:2383]: why do we need sudo here, and not on `systemctl`?
i would suggest to change the line 2386 to `self.remoter.sudo()` as IIRC `systemctl` also requires `sudo` permission

File: sdcm/nemesis.py [hunk 5]
```diff
@@ -389,6 +389,11 @@ def _kill_scylla_daemon(self):
             # Let's wait for the target Node to have their services re-started
             self.log.info('Waiting scylla services to be restarted after we killed them...')
             self.target_node.wait_db_up(timeout=14400)
+            if self.cluster.params.get('use_mgmt'):
+                # Workaround for https://github.com/scylladb/scylla-manager/issues/2813
+                # When scylla take too long time to bring api port up
+                #  scylla-manager-agent fails to start and never go up
+                self.target_node.start_service('scylla-manager-agent', timeout=600, ignore_status=True)
```
Comment (fgelcer) [inline sdcm/nemesis.py:396]: shouldn't it be `restart` instead of `start`?

Comment (dkropachev) [reply sdcm/nemesis.py:396]: fixed

File: sdcm/cluster.py (hunk 3 above)
Comment (bentsi) [reply sdcm/cluster.py:2386]: actually after looking in the `self.systemctl` we can't do it...

File: sdcm/nemesis.py (hunk 5 above)
Comment (dkropachev) [reply sdcm/nemesis.py:396]: Nope, if it is started we need to keep it.

File: sdcm/cluster.py (hunk 4 above)
Comment (fgelcer) [reply sdcm/cluster.py:2383]: i understand it now from @bentsi 's comments

Comment (bentsi) [reply sdcm/cluster.py:2383]: it is actually selected here: `sdcm.cluster.BaseNode.wrap_cmd_with_permission`

**Review by fgelcer — DISMISSED** [review]
LGTM

File: sdcm/cluster.py [hunk 6]
```diff
@@ -2377,23 +2377,34 @@ def add_install_prefix(self, abs_path):
...
+        else:
+            cmd = f'{self.systemctl} {cmd} {service_name}.service'
+        self.remoter.run(cmd, **opts)
+
+    def start_service(self, service_name: str, timeout: int = 500, ignore_status=False):
+        self._service_cmd(service_name, 'start', timeout=timeout, ignore_status=ignore_status)
```
Comment (bentsi) [inline sdcm/cluster.py:2389]: please use kw arguments

File: sdcm/cluster.py [hunk 7]
```diff
@@ -2377,23 +2377,34 @@ def add_install_prefix(self, abs_path):
         }
         return checklist.get(abs_path, INSTALL_DIR + abs_path)
 
+    def _service_cmd(self, service_name: str, cmd: str, timeout: int = 500, ignore_status=False):
+        opts = {'timeout': timeout, 'ignore_status': ignore_status}
```
Comment (bentsi) [inline sdcm/cluster.py:2381]: not needed

**Review by bentsi — APPROVED** [review]

**Pull request merged**